lib/HdtDocument.cc

```diff
@@ -368,6 +369,111 @@ NAN_METHOD(HdtDocument::SearchTerms) {
     info[4]->IsObject() ? info[4].As<Object>() : info.This()));
 }
 
+/******** HdtDocument#_fetchDistinctTerms ********/
+
+class FetchDistinctTermsWorker : public Nan::AsyncWorker {
+  HdtDocument* document;
+  // JavaScript function arguments
+  string filterVal;
+  hdt::TripleComponentRole filterPos;
+  uint32_t limit;
+  hdt::TripleComponentRole position;
+  Persistent<Object> self;
+  // Callback return values
+  vector<string> distinctTerms;
+public:
+  FetchDistinctTermsWorker(HdtDocument* document, char* filterVal, uint32_t filterPos,
+                    uint32_t limit, uint32_t posId, Nan::Callback* callback,
+                    Local<Object> self)
+    : Nan::AsyncWorker(callback),
+      document(document), filterVal(filterVal), filterPos((TripleComponentRole)  filterPos),
+      limit(limit), position((TripleComponentRole) posId) {
+        SaveToPersistent("self", self);
+      };
+
+  void Execute() {
+
+    const char *subject = "";
+    const char *predicate = "";
+    const char *object = "";
+
+    // Initialize filter search term.
+    switch (filterPos) {
+      case SUBJECT:
+        subject = filterVal.c_str();
+        break;
+      case PREDICATE:
+        predicate = filterVal.c_str();
+        break;
+      case OBJECT:
+        object = filterVal.c_str();
+        break;
+    };
```

Comment:
The above 16 lines only express a very simple idea, so should be made more succinct, such as:

```C++
const char *subject = filterPos === SUBJECT ? filterVal.c_str() : "";
```

Reply:
…but actually, can all of these cases occur? (See JS-land code.)

Reply:
Yes you are right. In the beginning I was thinking keeping things more generic, in case wish to add other combinations of positions when fetching distinct terms. 

Since we are interested in fetching distinct predicates for a given object and there are already some checks in JS that prevent these cases to occur, I will remove these lines.